src/main/java/emu/grasscutter/game/combine/CombineManger.java

```diff
@@ -87,11 +92,121 @@ public CombineResult combineItem(Player player, int cid, int count) {
         result.setMaterial(List.of());
         result.setResult(
                 List.of(
-                        new ItemParamData(
-                                combineData.getResultItemId(), combineData.getResultItemCount() * count)));
-        // TODO lucky characters
-        result.setExtra(List.of());
-        result.setBack(List.of());
+                        new ItemParamData(combineData.getResultItemId(), combineData.getResultItemCount())));
+        // lucky characters
+        int luckyCount = 0;
+        Avatar avatar = player.getAvatars().getAvatarByGuid(avatarGuid);
+        CombineBonusData combineBonusData = GameData.getCombineBonusDataMap().get(avatar.getAvatarId());
+        if (combineBonusData != null
+                && combineData.getCombineType() == combineBonusData.getCombineType()) {
+            double luckyChange = combineBonusData.getParamVec().get(0);
+            for (int i = 0; i < count; i++) {
+                if (ThreadLocalRandom.current().nextDouble() <= luckyChange) {
+                    luckyCount++;
+                }
+            }
+        }
+
+        result.setExtra(new ArrayList<ItemParamData>());
+        result.setBack(new ArrayList<ItemParamData>());
+        result.setRandom(new ArrayList<ItemParamData>());
+
+        // add lucky items
+        if (luckyCount > 0) {
+            switch (combineBonusData.getBonusType()) {
```

Comment:
instead of using the raw string here, consider making an enum with reasonable serialized names or just the game games